flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -77,37 +77,34 @@
 	 */
 	private final LinkedHashSet<InputGate> inputGatesWithData = new LinkedHashSet<>();
 
-	/** The total number of input channels across all unioned input gates. */
-	private final int totalNumberOfInputChannels;
+	/** Input channels across all unioned input gates. */
+	private final InputChannel[] inputChannels;
 
 	/**
 	 * A mapping from input gate to (logical) channel index offset. Valid channel indexes go from 0
 	 * (inclusive) to the total number of input channels (exclusive).
 	 */
-	private final Map<InputGate, Integer> inputGateToIndexOffsetMap;
+	private final int[] inputGateChannelIndexOffsets;
 
 	public UnionInputGate(InputGate... inputGates) {
 		this.inputGates = checkNotNull(inputGates);
 		checkArgument(inputGates.length > 1, "Union input gate should union at least two input gates.");
 
-		this.inputGateToIndexOffsetMap = Maps.newHashMapWithExpectedSize(inputGates.length);
 		this.inputGatesWithRemainingData = Sets.newHashSetWithExpectedSize(inputGates.length);
 
+		final int maxGateIndex = Arrays.stream(inputGates).mapToInt(InputGate::getGateIndex).max().orElse(0);
```

Comment:
I am wondering whether the `inputGates` argument might refer to `UnionInputGate` in practice, otherwise it should be definitely defined `SingleInputGate`. For the case of union gate in arguments, `getGateIndex` would throw exception as implemented default in `InputGate`.

Reply:
Added a check that no union of UnionInputGate are possible - it happened to late as you have observed.

Reply:
Changed it even more with an extra commit that introduces `IndexedInputGate`. Only this `InputGate` has an `getIndex` and union gate can only be composed by these gates (turning the runtime into a compile time check).

Reply:
From the `UnionInputGate` java doc:
>  * <strong>It is NOT possible to recursively union union input gates.</strong>